man/res_group.5
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "RES_GROUP 1"
.TH RES_GROUP 1 "2014-09-13" "Clockwork v2.3.0" "Clockwork Manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
res_group \- Clockwork Resource Type for system group accounts
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The group resource manages system group accounts.
.SH "ATTRIBUTES"
.IX Header "ATTRIBUTES"
.IP "\fBgid\fR" 4
.IX Item "gid"
Numeric group \s-1ID\s0 of this account.
.IP "\fBname\fR" 4
.IX Item "name"
Name of this group account.
.IP "\fBpresent\fR" 4
.IX Item "present"
Whether or not this account should exist on the system.
.IP "\fBmember\fR" 4
.IX Item "member"
Adds or removes a user account from the list of permanent group
members.
.Sp
If set to the name of a user, that user will be added
to the group if they do not already belong.
.Sp
If the username is prefixed with a \*(L"!\*(R" (as in, negation), that user
will be removed from the group if they currently belong to it.
.IP "\fBadmin\fR" 4
.IX Item "admin"
Adds or removes a user account from the list of group administrators.
.Sp
If set to the name of a user, that user will be added
to the group administrator list if they are not already on it.
.Sp
If the username is prefixed with a \*(L"!\*(R" (as in, negation), that user
will be removed from the group administrator list if they are
currently on it.
.IP "\fBmembers\fR" 4
.IX Item "members"
Like \fBmember\fR, except that it operates on a space-separated list
of usernames to add or remove.  Negation rules still apply.
.IP "\fBadmins\fR" 4
.IX Item "admins"
Like \fBadmin\fR, except that it takes a space-separated list
of usernames to add or remove.  Negation rules still apply.
.IP "\fBpassword\fR" 4
.IX Item "password"
.PD 0
.IP "\fBpwhash\fR" 4
.IX Item "pwhash"
.PD
The (encrypted) password for this account.  Group passwords are used
to allow non-members to assume temporary membership in a group, if they
know the password.
.Sp
Details of encryption depend on the specific platform and system
configuration.  Clockwork does not attempt to encrypt or decrypt passwords
for groups.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.SS "Basic Group Management"
.IX Subsection "Basic Group Management"
This example ensures that an \fIadmins\fR group exists on the system
with a \s-1GID\s0 of 42.  This type of record can be used in conjunction with
user resources (see \fBres_user\fR(5)) to define primary groups.
.PP
.Vb 3
\&    group "admins" {
\&      gid: 42
\&    }
.Ve
.SS "Member Management"
.IX Subsection "Member Management"
In this example, the policy ensures that the \*(L"projectx\*(R" group exists,
that the users \fIjose\fR, \fIbob\fR and \fIeva\fR belong to the group,
and that \fIalice\fR does not.
.PP
.Vb 3
\&    group "projectx" {
\&      gid:      1337
\&      password: "!!"
\&
\&      # add these users
\&      member:   "jose"
\&      member:   "bob"
\&      member:   "eva"
\&
\&      # remove these (note the \*(Aq!\*(Aq)
\&      member:   "!alice"
\&    }
.Ve
.PP
The \fBpassword\fR attribute is specified as an invalid hash to ensure
membership in the projectx group is managed strictly by the system
administrators (and, by extension, Clockwork).
.PP
Another way to handle the membership list is via \fBmembers\fR,
like this:
.PP
.Vb 5
\&    group "projectx" {
\&      gid:      1337
\&      password: "!!"
\&      members:  "jose bob eva !alice"
\&    }
.Ve
.SS "Administrator Management"
.IX Subsection "Administrator Management"
Management of group administrators works just like group membership
management.  To ensure that \fIjames\fR is an admin in the \*(L"projectx\*(R"
group, but that \fInick\fR is not:
.PP
.Vb 5
\&    group "projectx" {
\&      gid:   1337
\&      admin: "james"
\&      admin: "!nick"
\&    }
.Ve
.SH "CAVEATS"
.IX Header "CAVEATS"
.SS "1. Primary vs. Auxiliary Groups"
.IX Subsection "1. Primary vs. Auxiliary Groups"
Under conventional user and group mechanics, a user's primary group
is not included in their list of auxiliary groups.  To illustrate,
the following policy snippet does not affect \fIjoe\fR's primary
group:
.PP
.Vb 4
\&    group "users" {
\&        gid:    1234
\&        member: "!joe"
\&    }
\&
\&    user "joe" {
\&        uid:  100
\&        gid:  1234
\&    }
.Ve
.PP
\&\fIjoe\fR's primary group will still be \fIusers\fR, even through
the group definition attempts to remove him.
.SS "2. Duplicate GIDs"
.IX Subsection "2. Duplicate GIDs"
Clockwork does not attempt to enforce the uniqueness of group GIDs
on the local system.
.SH "DEPENDENCIES"
.IX Header "DEPENDENCIES"
None.
.SH "AUTHOR"
.IX Header "AUTHOR"
Clockwork was designed and written by James Hunt.


CHANGELOG
Release Notes for Clockwork


v2.3.0 - 2014-09-11

  * New `cw' multi-call binary to replace cw* tools into a single,
    user-extensible sub-command dispather, a la git.

    The following utilities have therefore been renamed:

       cwcc     ->  cw-cc
       fact     ->  cw-fact
       cwtrust  ->  cw-trust
       cwpol    ->  cw-shell
       cwkey    ->  cw-cert
       cw-run   ->  cw-mesh (and cw-show, cw-ping, etc.)

       cwtool   ->  cw-localsys [internal to cogd]

   * cw-mesh prompting is now done to stderr, for pipelines
   * cw-mesh now ignores nodes that optout by default
     (use --optout to retain the legacy behavior)
   * cw-mesh now honors ~/.cwrc as an adjunct to /etc/clockwork/cw.conf




res_group.c
#include <assert.h>
#include <string.h>
#include <errno.h>

#include "res_group.h"
#include "pack.h"


#define RES_GROUP_PACK_PREFIX "res_group::"
#define RES_GROUP_PACK_OFFSET 11
/* Pack format for res_group structure:
     L - rg_enf
     a - rg_name
     a - rg_passwd
     L - rg_gid
     a - rg_mem_add (joined stringlist)
     a - rg_mem_rm  (joined stringlist)
     a - rg_adm_add (joined stringlist)
     a - rg_adm_rm  (joined stringlist)
 */
#define RES_GROUP_PACK_FORMAT "LaaLaaaa"


static int _res_group_diff(struct res_group*);
static int _group_update(stringlist*, stringlist*, const char*);

/*****************************************************************/

static int _res_group_diff(struct res_group *rg)
{
	assert(rg);
	assert(rg->rg_grp);
	assert(rg->rg_sg);

	stringlist *list; /* for diff'ing gr_mem and sg_adm */

	rg->rg_diff = RES_GROUP_NONE;

	if (res_group_enforced(rg, NAME) && strcmp(rg->rg_name, rg->rg_grp->gr_name) != 0) {
		rg->rg_diff |= RES_GROUP_NAME;
	}

	if (res_group_enforced(rg, PASSWD) && strcmp(rg->rg_passwd, rg->rg_sg->sg_passwd) != 0) {
		rg->rg_diff |= RES_GROUP_PASSWD;
	}

	if (res_group_enforced(rg, GID) && rg->rg_gid != rg->rg_grp->gr_gid) {
		rg->rg_diff |= RES_GROUP_GID;
	}

	if (res_group_enforced(rg, MEMBERS)) {
		/* use list as a stringlist of gr_mem */
		list = stringlist_new(rg->rg_grp->gr_mem);
		if (stringlist_diff(list, rg->rg_mem) == 0) {
			rg->rg_diff |= RES_GROUP_MEMBERS;
		}
		stringlist_free(list);
	}

	if (res_group_enforced(rg, ADMINS)) {
		/* use list as a stringlist of sg_adm */
		list = stringlist_new(rg->rg_sg->sg_adm);
		if (stringlist_diff(list, rg->rg_adm) == 0) {
			rg->rg_diff |= RES_GROUP_ADMINS;
		}
		stringlist_free(list);
	}

	return 0;
}

static int _group_update(stringlist *add, stringlist *rm, const char *user)
{
	/* put user in add */
	if (stringlist_search(add, user) != 0) {
		stringlist_add(add, user);
	}

	/* take user out of rm */
	if (stringlist_search(rm, user) == 0) {
		stringlist_remove(rm, user);
	}

	return 0;
}

/*****************************************************************/

void* res_group_new(const char *key)
{
	struct res_group *rg;

	rg = xmalloc(sizeof(struct res_group));
	list_init(&rg->res);

	rg->rg_name = NULL;
	rg->rg_passwd = NULL;
	rg->rg_gid = 0;

	rg->rg_grp = NULL;
	rg->rg_sg  = NULL;

	rg->rg_mem = NULL;
	rg->rg_mem_add = stringlist_new(NULL);
	rg->rg_mem_rm  = stringlist_new(NULL);

	rg->rg_adm = NULL;
	rg->rg_adm_add = stringlist_new(NULL);
	rg->rg_adm_rm  = stringlist_new(NULL);

	rg->rg_enf  = RES_GROUP_NONE;
	rg->rg_diff = RES_GROUP_NONE;

	if (key) {
		res_group_set(rg, "name", key);
		rg->key = string("res_group:%s", key);
	} else {
		rg->key = NULL;
	}

	return rg;
}

void res_group_free(void *res)
{
	struct res_group *rg = (struct res_group*)(res);

	if (rg) {
		free(rg->key);

		free(rg->rg_name);
		free(rg->rg_passwd);

		list_del(&rg->res);

		if (rg->rg_mem) {
			stringlist_free(rg->rg_mem);
		}
		stringlist_free(rg->rg_mem_add);
		stringlist_free(rg->rg_mem_rm);

		if (rg->rg_adm) {
			stringlist_free(rg->rg_adm);
		}
		stringlist_free(rg->rg_adm_add);
		stringlist_free(rg->rg_adm_rm);
	}
	free(rg);
}

const char* res_group_key(const void *res)
{
	const struct res_group *rg = (struct res_group*)(res);
	assert(rg);

	return rg->key;
}

int res_group_norm(void *res) { return 0; }

int res_group_set(void *res, const char *name, const char *value)
{
	struct res_group *rg = (struct res_group*)(res);
	assert(res);

	if (strcmp(name, "gid") == 0) {
		rg->rg_gid = strtoll(value, NULL, 10);
		rg->rg_enf |= RES_GROUP_GID;

	} else if (strcmp(name, "name") == 0) {
		free(rg->rg_name);
		rg->rg_name = strdup(value);
		rg->rg_enf |= RES_GROUP_NAME;

	} else if (strcmp(name, "present") == 0) {
		if (strcmp(value, "no") != 0) {
			rg->rg_enf ^= RES_GROUP_ABSENT;
		} else {
			rg->rg_enf |= RES_GROUP_ABSENT;
		}

	} else if (strcmp(name, "member") == 0) {
		if (value[0] == '!') {
			return res_group_remove_member(rg, value+1);
		} else {
			return res_group_add_member(rg, value);
		}
	} else if (strcmp(name, "admin") == 0) {
		if (value[0] == '!') {
			return res_group_remove_admin(rg, value+1);
		} else {
			return res_group_add_admin(rg, value);
		}
	} else if (strcmp(name, "pwhash") == 0 || strcmp(name, "password") == 0) {
		free(rg->rg_passwd);
		rg->rg_passwd = strdup(value);
		rg->rg_enf |= RES_GROUP_PASSWD;

	} else {
		return -1;
	}

	return 0;
}

int res_group_enforce_members(struct res_group *rg, int enforce)
{
	assert(rg);

	if (enforce) {
		rg->rg_enf |= RES_GROUP_MEMBERS;
	} else {
		rg->rg_enf ^= RES_GROUP_MEMBERS;
	}
	return 0;
}

/* updates rg_mem_add */
int res_group_add_member(struct res_group *rg, const char *user)
{
	assert(rg);
	assert(user);

	res_group_enforce_members(rg, 1);
	/* add to rg_mem_add, remove from rg_mem_rm */
	return _group_update(rg->rg_mem_add, rg->rg_mem_rm, user);
}
/* updates rg_mem_rm */
int res_group_remove_member(struct res_group *rg, const char *user)
{
	assert(rg);
	assert(user);

	res_group_enforce_members(rg, 1);
	/* add to rg_mem_rm, remove from rg_mem_add */
	return _group_update(rg->rg_mem_rm, rg->rg_mem_add, user);
}

int res_group_enforce_admins(struct res_group *rg, int enforce)
{
	assert(rg);

	if (enforce) {
		rg->rg_enf |= RES_GROUP_ADMINS;
	} else {
		rg->rg_enf ^= RES_GROUP_ADMINS;
	}
	return 0;
}

/* updates rg_adm_add */
int res_group_add_admin(struct res_group *rg, const char *user)
{
	assert(rg);
	assert(user);

	res_group_enforce_admins(rg, 1);
	/* add to rg_adm_add, remove from rg_adm_rm */
	return _group_update(rg->rg_adm_add, rg->rg_adm_rm, user);
}

/* updates rg_adm_rm */
int res_group_remove_admin(struct res_group *rg, const char *user)
{
	assert(rg);
	assert(user);

	res_group_enforce_admins(rg, 1);
	/* add to rg_adm_rm, remove from rg_adm_add */
	return _group_update(rg->rg_adm_rm, rg->rg_adm_add, user);
}

int res_group_stat(void *res, const struct resource_env *env)
{
	struct res_group *rg = (struct res_group*)(res);
	assert(rg);
	assert(env);
	assert(env->group_grdb);
	assert(env->group_sgdb);

	rg->rg_grp = grdb_get_by_name(env->group_grdb, rg->rg_name);
	rg->rg_sg = sgdb_get_by_name(env->group_sgdb, rg->rg_name);
	if (!rg->rg_grp || !rg->rg_sg) { /* new group */
		rg->rg_diff = rg->rg_enf;

		rg->rg_mem = stringlist_new(NULL);
		stringlist_add_all(rg->rg_mem, rg->rg_mem_add);

		rg->rg_adm = stringlist_new(NULL);
		stringlist_add_all(rg->rg_adm, rg->rg_adm_add);

		return 0;
	}

	/* set up rg_mem as the list we want for gr_mem */
	rg->rg_mem = stringlist_new(rg->rg_grp->gr_mem);
	stringlist_add_all(rg->rg_mem, rg->rg_mem_add);
	stringlist_remove_all(rg->rg_mem, rg->rg_mem_rm);
	stringlist_uniq(rg->rg_mem);

	/* set up rg_adm as the list we want for sg_adm */
	rg->rg_adm = stringlist_new(rg->rg_sg->sg_adm);
	stringlist_add_all(rg->rg_adm, rg->rg_adm_add);
	stringlist_remove_all(rg->rg_adm, rg->rg_adm_rm);
	stringlist_uniq(rg->rg_adm);

	return _res_group_diff(rg);
}

struct report* res_group_fixup(void *res, int dryrun, const struct resource_env *env)
{
	struct res_group *rg = (struct res_group*)(res);
	assert(rg);
	assert(env);
	assert(env->group_grdb);
	assert(env->group_sgdb);

	struct report *report;
	char *action;

	int new_group;
	stringlist *orig;
	stringlist *to_add;
	stringlist *to_remove;

	report = report_new("Group", rg->rg_name);

	/* Remove the group if RES_GROUP_ABSENT */
	if (res_group_enforced(rg, ABSENT)) {
		if (rg->rg_grp || rg->rg_sg) {
			action = string("remove group");

			if (dryrun) {
				report_action(report, action, ACTION_SKIPPED);
			} else {
				if ((rg->rg_grp && grdb_rm(env->group_grdb, rg->rg_grp) != 0)
				 || (rg->rg_sg && sgdb_rm(env->group_sgdb, rg->rg_sg) != 0)) {
					report_action(report, action, ACTION_FAILED);
				} else {
					report_action(report, action, ACTION_SUCCEEDED);
				}
			}
		}

		return report;
	}

	if (!rg->rg_grp || !rg->rg_sg) {
		new_group = 1;
		action = string("create group");

		if (!rg->rg_grp) { rg->rg_grp = grdb_new_entry(env->group_grdb, rg->rg_name, rg->rg_gid); }
		if (!rg->rg_sg)  { rg->rg_sg = sgdb_new_entry(env->group_sgdb, rg->rg_name); }

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else if (rg->rg_grp && rg->rg_sg) {
			report_action(report, action, ACTION_SUCCEEDED);
		} else {
			report_action(report, action, ACTION_FAILED);
		}
	}

	if (res_group_different(rg, PASSWD)) {
		if (new_group) {
			action = string("set group membership password");
		} else {
			action = string("change group membership password");
		}

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else {
			xfree(rg->rg_grp->gr_passwd);
			rg->rg_grp->gr_passwd = strdup("x");
			xfree(rg->rg_sg->sg_passwd);
			rg->rg_sg->sg_passwd = strdup(rg->rg_passwd);

			report_action(report, action, ACTION_SUCCEEDED);
		}
	}

	if (res_group_different(rg, GID)) {
		if (new_group) {
			action = string("set gid to %u", rg->rg_gid);
		} else {
			action = string("change gid from %u to %u", rg->rg_grp->gr_gid, rg->rg_gid);
		}

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else {
			rg->rg_grp->gr_gid = rg->rg_gid;
			report_action(report, action, ACTION_SUCCEEDED);
		}
	}

	if (res_group_enforced(rg, MEMBERS) && res_group_different(rg, MEMBERS)) {
		orig = stringlist_new(rg->rg_grp->gr_mem);
		to_add = stringlist_dup(rg->rg_mem_add);
		stringlist_remove_all(to_add, orig);

		to_remove = stringlist_intersect(orig, rg->rg_mem_rm);

		if (!dryrun) {
			/* replace gr_mem and sg_mem with the rg_mem stringlist */
			xarrfree(rg->rg_grp->gr_mem);
			rg->rg_grp->gr_mem = xarrdup(rg->rg_mem->strings);

			xarrfree(rg->rg_sg->sg_mem);
			rg->rg_sg->sg_mem = xarrdup(rg->rg_mem->strings);
		}

		size_t i;
		for (i = 0; i < to_add->num; i++) {
			report_action(report, string("add %s", to_add->strings[i]), (dryrun ? ACTION_SKIPPED : ACTION_SUCCEEDED));
		}
		for (i = 0; i < to_remove->num; i++) {
			report_action(report, string("remove %s", to_remove->strings[i]), (dryrun ? ACTION_SKIPPED : ACTION_SUCCEEDED));
		}

		stringlist_free(orig);
		stringlist_free(to_add);
		stringlist_free(to_remove);
	}

	if (res_group_enforced(rg, ADMINS) && res_group_different(rg, ADMINS)) {
		orig = stringlist_new(rg->rg_sg->sg_adm);
		to_add = stringlist_dup(rg->rg_adm_add);
		stringlist_remove_all(to_add, orig);

		to_remove = stringlist_intersect(orig, rg->rg_mem_add);

		if (!dryrun) {
			/* replace sg_adm with the rg_adm stringlist */
			xarrfree(rg->rg_sg->sg_adm);
			rg->rg_sg->sg_adm = xarrdup(rg->rg_adm->strings);
		}

		size_t i;
		for (i = 0; i < to_add->num; i++) {
			report_action(report, string("grant admin rights to %s", to_add->strings[i]), (dryrun ? ACTION_SKIPPED : ACTION_SUCCEEDED));
		}
		for (i = 0; i < to_remove->num; i++) {
			report_action(report, string("revoke admin rights from %s", to_remove->strings[i]), (dryrun ? ACTION_SKIPPED : ACTION_SUCCEEDED));
		}

		stringlist_free(orig);
		stringlist_free(to_add);
		stringlist_free(to_remove);
	}

	return report;
}

int res_group_is_pack(const char *packed)
{
	return strncmp(packed, RES_GROUP_PACK_PREFIX, RES_GROUP_PACK_OFFSET);
}

char *res_group_pack(const void *res)
{
	const struct res_group *rg = (const struct res_group*)(res);
	assert(rg);

	char *tmp;
	char *mem_add = NULL, *mem_rm = NULL,
	     *adm_add = NULL, *adm_rm = NULL;

	if (!(mem_add = stringlist_join(rg->rg_mem_add, "."))
	 || !(mem_rm  = stringlist_join(rg->rg_mem_rm,  "."))
	 || !(adm_add = stringlist_join(rg->rg_adm_add, "."))
	 || !(adm_rm  = stringlist_join(rg->rg_adm_rm,  "."))) {
		free(mem_add); free(mem_rm);
		free(adm_add); free(adm_rm);
		return NULL;
	}

	tmp = pack(RES_GROUP_PACK_PREFIX, RES_GROUP_PACK_FORMAT,
	           rg->rg_enf,
	           rg->rg_name, rg->rg_passwd, rg->rg_gid,
	           mem_add, mem_rm, adm_add, adm_rm);

	free(mem_add); free(mem_rm);
	free(adm_add); free(adm_rm);

	return tmp;
}

void* res_group_unpack(const char *packed)
{
	char *mem_add = NULL, *mem_rm = NULL,
	     *adm_add = NULL, *adm_rm = NULL;
	struct res_group *rg = res_group_new(NULL);

	if (unpack(packed, RES_GROUP_PACK_PREFIX, RES_GROUP_PACK_FORMAT,
		&rg->rg_enf,
		&rg->rg_name, &rg->rg_passwd, &rg->rg_gid,
		&mem_add, &mem_rm, &adm_add, &adm_rm)) {

		res_group_free(rg);
		return NULL;
	}

	rg->key = string("res_group:%s", rg->rg_name);

	stringlist_free(rg->rg_mem_add);
	rg->rg_mem_add = stringlist_split(mem_add, strlen(mem_add), ".");

	stringlist_free(rg->rg_mem_rm);
	rg->rg_mem_rm  = stringlist_split(mem_rm,  strlen(mem_rm),  ".");

	stringlist_free(rg->rg_adm_add);
	rg->rg_adm_add = stringlist_split(adm_add, strlen(adm_add), ".");

	stringlist_free(rg->rg_adm_rm);
	rg->rg_adm_rm  = stringlist_split(adm_rm,  strlen(adm_rm),  ".");

	free(mem_add); free(mem_rm);
	free(adm_add); free(adm_rm);

	return rg;
}
